sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/DataflowGraph.scala

```diff
@@ -0,0 +1,241 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.graph
+
+import scala.util.Try
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.pipelines.graph.DataflowGraph.mapUnique
+import org.apache.spark.sql.pipelines.util.SchemaMergingUtils
+import org.apache.spark.sql.types.StructType
+
+/**
+ * DataflowGraph represents the core graph structure for Spark declarative pipelines.
+ * It manages the relationships between logical flows, tables, and views, providing
+ * operations for graph traversal, validation, and transformation.
+ */
+case class DataflowGraph(flows: Seq[Flow], tables: Seq[Table], views: Seq[View])
+    extends GraphOperations
+    with GraphValidations {
+
+  /** Returns a [[Output]] given its identifier */
+  lazy val output: Map[TableIdentifier, Output] = mapUnique(tables, "output")(_.identifier)
+
+  /**
+   * Returns [[Flow]]s in this graph that need to get planned and potentially executed when
+   * executing the graph. Flows that write to logical views are excluded.
+   */
+  lazy val materializedFlows: Seq[ResolvedFlow] = {
+    resolvedFlows.filter(
+      f => output.contains(f.destinationIdentifier)
+    )
+  }
+
+  /** Returns the identifiers of [[materializedFlows]]. */
+  val materializedFlowIdentifiers: Set[TableIdentifier] = materializedFlows.map(_.identifier).toSet
+
+  /** Returns a [[Table]] given its identifier */
```

Comment:
The comment is odd. It is a map from table id to Table, instead of a method which returns a Table.
There are many similar comments in this file.